design/src/rfcs/rfc0024_supporting_select_object_content_scan_range.md

```diff
@@ -0,0 +1,214 @@
+```
+
+Additionally, the `@default` trait doesn't _(and likely will never)_ support the ability to define a dynamic value like
+the _"one less than the size of the object being queried"_ behavior that the docs say `end` is meant to default to.
+
+**Once this RFC is implemented,** users will finally be able to use the `ScanRange` input as documented:
+- Default values will be equal to those specified in the `ScanRange` docs.
+    - For `start`, the default value is zero.
+    - For `end`, the default value is one less than the size of the object being queried.
+- If `start` is set and `end` is "unset", the object will be scanned from `start` to then end of the file.
+- If `end` is set and `start` is "unset", the last `N` bytes of the object will be read, where `N == end`.
+- If both fields are set, the object will be scanned from `start` to `end`
+- If both fields are "unset", an error will be returned because at least one must be set.
+
+How to actually implement this RFC
+----------------------------------
+
+In order to implement this feature, we would extend the current S3 customization to prevent `ScanRange` from being
+generated normally and would instead insert inlineable code into `aws_sdk_s3::models` that provided a handwritten
+version of `ScanRange` with special behavior. We would generate a serialization function as well that handles
+defaults and "unset" fields as defined in the `ScanRange` docs.
+
+By generating an enum instead of a struct, we could allow users to be explicit about the behavior they expect.
+
+```rust
+#[non_exhaustive]
+#[derive(Clone, PartialEq)]
+pub enum ScanRange {
+    StartToEnd {
+        start: i64,
+        end: i64,
+    },
+    StartToEndOfObject {
+        start: i64,
+    },
+    LastNBytesOfObject {
+      n_bytes: i64
+    }
+}
+
+impl ScanRange {
+    pub fn start_to_end(start: i64, end: i64) -> Self {
+        ScanRange::StartToEnd { start, end }
+    }
+
+    pub fn start_to_end_of_object(start: i64) -> Self {
+        ScanRange::StartToEndOfObject { start }
+    }
+
+    pub fn last_n_bytes_of_object(n_bytes: i64) -> Self {
+      ScanRange::LastNBytesOfObject { n_bytes }
+    }
+
+    pub fn start(&self) -> Option<i64> {
+        match self {
+            ScanRange::StartToEnd { start, .. } => Some(start),
+            ScanRange::StartToEndOfObject { start } => Some(start),
+            ScanRange::LastNBytesOfObject { .. } => None,
+        }
+    }
+
+    pub fn end(&self) -> Option<i64> {
+        match self {
+            ScanRange::StartToEnd { end, .. } => Some(end),
+            ScanRange::StartToEndOfObject { .. } => None,
+            ScanRange::LastNBytesOfObject { n_bytes } => Some(n_bytes),
+        }
+    }
+}
+```
+
+To ensure that this handwritten struct doesn't deviate from the S3 model, we'd create a codegen test that would verify
+compatibility of the two models.
+
+Why this RFC should be rejected
+-------------------------------
+
+There are several reasons to reject this RFC:
```

Comment:
Another reason: In the unlikely event that S3 modifies `ScanRange`, the handwritten implementation will be out of sync.